Add `read_buf` equivalents for positioned reads

niklasf opened the pull request:
Adds the following items under the ~~`read_buf` (rust-lang/rust#78485)~~ `read_buf_at` (rust-lang/rust#140771) feature:

 - `std::os::unix::fs::FileExt::read_buf_at`
 - `std::os::unix::fs::FileExt::read_buf_exact_at`
 - `std::os::windows::fs::FileExt::seek_read_buf`

try-job: `x86_64-msvc*`
try-job: `test-various*`
try-job: `dist-various*`

niklasf committed:
Add `read_buf` equivalents for positioned reads

Adds the following items under the `read_buf_at` feature:

 - `std::os::unix::fs::FileExt::read_buf_at`
 - `std::os::unix::fs::FileExt::read_buf_exact_at`
 - `std::os::windows::fs::FileExt::seek_read_buf`

tgross35 commented:
Have these changes been discussed with libs-api at all? Usually changes to unstable API need to be proposed at https://github.com/rust-lang/libs-team/ with the ACP issue template.

This needs some tests as well.

tgross35 commented:
For the above,
@rustbot author

Kivooeo commented:
For consistency and clarity, could we include safety comments on each unsafe block? Even brief notes are helpful to understand the assumptions being made

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14983261821/job/42092086795) [(plain)](https://github.com/rust-lang/rust/commit/09ac65a8551eb5710acb87447ecf8f6969c42923/checks/42092086795/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14983563070/job/42093044163) [(plain)](https://github.com/rust-lang/rust/commit/a22d2b4808172b5a050eebf4a3b94863339c9993/checks/42093044163/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14983920509/job/42094097289) [(plain)](https://github.com/rust-lang/rust/commit/59cb58ebfc2f7cf39219e5c0918798618a608fe4/checks/42094097289/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

niklasf commented:
Thank you both.

* Opened an APC that was accepted, but items to be added under a new feature instead of `read_buf`. Moved to `read_buf_at` with new tracking issue #140771.
* Added safety comments also for FFI calls.
* Added tests.

@rustbot ready

tgross35 commented on library/std/src/os/unix/fs.rs:
The newly added API needs some doc examples

tgross35 commented on library/std/src/sys/fd/unix.rs:
```diff
@@ -87,6 +87,19 @@ const fn max_iov() -> usize {
     16 // The minimum value required by POSIX.
 }
 
+#[cfg(not(any(
+    all(target_os = "linux", not(target_env = "musl")),
+    target_os = "android",
+    target_os = "hurd"
+)))]
+use libc::pread as pread64;
+#[cfg(any(
+    all(target_os = "linux", not(target_env = "musl")),
+    target_os = "android",
+    target_os = "hurd"
+))]
+use libc::pread64;
```
Since this is getting moved anyway, mind simplifying this with `cfg_select!`?

niklasf replied:
Sure. Did I pick the right place for `#![feature(cfg_select)]`?

tgross35 commented:
@bors try

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15312404166/job/43083910426)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/15312404166/job/43083910426","builder_name":"checks-actions"} -->

niklasf commented:
Ready for the next round of review.

The failed test on Windows was an assertion that I had (blindly) added about the seek position after `seek_read_buf()` with an empty buffer. Turns out it's a noop, but neither the existing `seek_read()` nor the current underlying Windows API guarantees it (afaict), so I just removed the assertion.

@rustbot ready

tgross35 commented:
@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #145334) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 reviewed (changes requested):
Thanks for the updates, two small nits then r=me

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/140459/aaf87a6d289053246d19f94fbc9972582816e93d..bf491a1ddea28c4c137e29ddfd633b87e8e3029f)*

niklasf commented on library/std/src/sys/fd/unix.rs:
```diff
@@ -369,7 +389,6 @@ impl FileDesc {
         )))
     }
 
-    #[cfg_attr(target_os = "vxworks", allow(unused_unsafe))]
     pub fn write_at(&self, buf: &[u8], offset: u64) -> io::Result<usize> {
```
After https://github.com/rust-lang/libc/pull/3727 this is no longer needed. So I didn't add `#[allow(unused_unsafe)]` to `read_buf_at()` and removed it everywhere else as a drive-by cleanup.

niklasf commented:
@rustbot ready

bors commented:
:pushpin: Commit c914c471b9d83f9e6c74b0e6c2c8044efd3a57dd has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 c914c471b9d83f9e6c74b0e6c2c8044efd3a57dd -->
<!-- homu: {"type":"Approved","sha":"c914c471b9d83f9e6c74b0e6c2c8044efd3a57dd","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
